### PR TITLE
[Breaking Change] Multiple port support for `GameServer`

### PR DESCRIPTION
**Pull request by markmandel**
This allows the `GameServer` (and anything that uses it as a template, such as `Fleets`), to provide configuration and assignment of multiple ports to the game server container.

For example:

```yaml
apiVersion: "stable.agones.dev/v1alpha1"
kind: GameServer
metadata:
  name: "simple-udp"
spec:
  ports:
  - name: default
    portPolicy: "dynamic"
    containerPort: 7654
  template:
    spec:
      containers:
      - name: simple-udp
        image: gcr.io/agones-images/udp-server:0.1
```

This also includes update to documentation - but in a separate bottom section of each document, to allow for moving into the live version at release time. (This is becoming a mess - a website can't come soon enough).

This also adjusts the `GameServer > Status`, to be able to handle multiple ports. It now looks like this:

```
Status:
  Address:    192.168.99.100
  Node Name:  agones
  Ports:
    Name:  default
    Port:  7502
  State:   Ready
```

While this a breaking change (due to the change in `GameServer > status`, the previous legacy configuration of still works (even thought it's no longer documented), as it is automatically converted to the new version.

Plans are to remove this legacy conversion functionality in 0.4.0, since being in alpha, we have a very low commitment to backward compatibility.

As an aside - versioning for CRDs is coming soon, so this will also solve this type of problem long term.

### CODE REVIEW

File: pkg/gameservers/portallocator.go [hunk 1]
```diff
@@ -156,33 +156,57 @@ func (pa *PortAllocator) syncPorts(key string) error {
...
+
+	type pn struct {
+		pa   portAllocation
+		port int32
+	}
+
+	// we only want this to be called inside the mutex lock
+	// so let's define the function here so it can never be called elsewhere.
+	// Also the return gives an escape from the double loop
+	findOpenPorts := func(amount int) []pn {
```
Comment (markmandel) [inline pkg/gameservers/portallocator.go]: @enocom would love to get your take on what I'm doing here.

Comment (enocom) [reply pkg/gameservers/portallocator.go]: I've seen this pattern of declaring a function inside a method and then used in the stdlib. It's something I've been meaning to do more often myself. Very nice.

File: examples/simple-udp/server/gameserver-legacy.yaml [hunk 2]
```diff
@@ -1,4 +1,4 @@
-# Copyright 2018 Google Inc. All Rights Reserved.
+# Copyright 2017 Google Inc. All Rights Reserved.
```
Comment (Kuqd) [inline examples/simple-udp/server/gameserver-legacy.yaml]: Back in past, Marty !

Comment (markmandel) [reply examples/simple-udp/server/gameserver-legacy.yaml]: Whoops. Will change.

File: pkg/apis/stable/v1alpha1/gameserver.go [hunk 3]
```diff
@@ -108,18 +110,10 @@ type GameServerSpec struct {
...
-	// When `static` is the policy specified, `HostPort` is required, to specify the port that game clients will
-	// connect to
-	PortPolicy PortPolicy `json:"portPolicy,omitempty"`
-	// ContainerPort is the port that is being opened on the game server process
-	ContainerPort int32 `json:"containerPort"`
-	// HostPort the port exposed on the host for clients to connect to
-	HostPort int32 `json:"hostPort,omitempty"`
-	// Protocol is the network protocol being used. Defaults to UDP. TCP is the only other option
-	Protocol corev1.Protocol `json:"protocol,omitempty"`
+	// GameServerPort is deprecated, to be removed in 0.4.0:
```
[the remark this answers is not in the document]
Comment (markmandel) [reply pkg/apis/stable/v1alpha1/gameserver.go]: When you say "all of this" - do you mean:
- drop the configuration backward compatibility for this release?
- Remove the comments on deprecation?
- Something else I'm not sure of?

[the remark this answers is not in the document]
Comment (Kuqd) [reply pkg/apis/stable/v1alpha1/gameserver.go]: yes the two first point :)

[the remark this answers is not in the document]
Comment (markmandel) [reply pkg/apis/stable/v1alpha1/gameserver.go]: I was just being nice giving people a short opportunity to migrate -- technically we are in alpha, so we can make breaking changes if need be, but I know people are actively using Agones in a couple of places, so figured I would be more gradual about breaking things :smile:  ?

Do you feel strongly that we should remove it?

[the remark this answers is not in the document]
Comment (markmandel) [reply pkg/apis/stable/v1alpha1/gameserver.go]: We had a chat - there was confusion around which release we were in. Looks like we're on the same page now, so merging this.

**Pull request merged**